File: companions/android/app/src/main/java/io/github/meyer4/secondlook/SafetyEngine.java
package io.github.meyer4.secondlook;

import org.json.*;
import java.net.URI;
import java.net.IDN;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.regex.*;

/** Offline, bounded, conservative checks. No Android services or network APIs. */
public final class SafetyEngine {
    public static final int MAX_TEXT = 12000, MAX_URL = 4096, MAX_LINKS = 10;
    public static final class Signal {
        public final String id, severity, title, detail;
        public Signal(String id, String severity, String title, String detail) {
            this.id=id; this.severity=severity; this.title=title; this.detail=detail;
        }
    }
    public static final class Result {
        public final String kind;
        public String hostname="";
        public boolean assumedHttps=false;
        public final List<Signal> signals=new ArrayList<>();
        public Result(String kind) { this.kind=kind; }
        public String level() {
            int medium=0;
            for(Signal signal:signals) { if(signal.severity.equals("high")) return "high"; if(signal.severity.equals("medium"))medium++; }
            return medium>=3 ? "high" : signals.isEmpty() ? "unknown" : "caution";
        }
        public String label() { return level().equals("high") ? "Strong warning signs" : level().equals("caution") ? "Worth a closer look" : "No common warning signs found"; }
        public String explanation() {
            return level().equals("high") ? "Pause before you click, pay, or share details. Verify the request through a channel you already trust. These patterns are not proof of a scam." : level().equals("caution") ? "Some details deserve a second look, but can also appear in legitimate content. Independently verify the request." : "The known patterns were not found. This does not mean the message or link is safe. Verify unexpected requests before acting.";
        }
    }
    private static final class Rule {
        final String id,severity,title,detail; final Pattern pattern; final boolean skipNegated;
        Rule(JSONObject json) throws JSONException {
            id=json.getString("id"); severity=json.getString("severity");title=json.getString("title");detail=json.getString("detail");
            pattern=Pattern.compile(json.getString("pattern"),Pattern.CASE_INSENSITIVE|Pattern.UNICODE_CASE);
            skipNegated=json.optBoolean("skipNegated",false);
        }
    }
    private final List<Rule> rules=new ArrayList<>();
    private final Set<String> shorteners=new HashSet<>();
    private final JSONArray brands;
    public final JSONArray playbook;
    private static final Pattern NEGATION=Pattern.compile("\\b(?:never|do\\s+not|don['’]t|should\\s+not|shouldn['’]t|will\\s+not|won['’]t|avoid)\\b[^,;.!?\\n]{0,65}$",Pattern.CASE_INSENSITIVE);
    private static final Pattern LINKS=Pattern.compile("(?:https?://|ftp://|javascript:|data:|file:///|www\\.)[^\\s<>\\\"`]+|\\b(?:[a-z\\d](?:[a-z\\d-]{0,61}[a-z\\d])?\\.)+[a-z]{2,63}(?:[/:?#][^\\s<>\\\"`]*)?",Pattern.CASE_INSENSITIVE);
    private static final Pattern HIDDEN=Pattern.compile("[\\u200B-\\u200F\\u202A-\\u202E\\u2066-\\u2069\\uFEFF]");
    public SafetyEngine(String json) throws JSONException {
        JSONObject data=new JSONObject(json);JSONArray source=data.getJSONArray("messageRules");
        for(int i=0;i<source.length();i++)rules.add(new Rule(source.getJSONObject(i)));
        JSONArray shorts=data.getJSONArray("shorteners");for(int i=0;i<shorts.length();i++)shorteners.add(shorts.getString(i));
        brands=data.getJSONArray("brands");playbook=data.getJSONArray("playbook");
    }
    private static boolean belongs(String host,String domain) { return host.equals(domain)||host.endsWith("."+domain); }
    private static void add(Result result,String id,String severity,String title,String detail) { result.signals.add(new Signal(id,severity,title,detail)); }
    public Result checkMessage(String input) {
        String text=input==null?"":input.trim();
        if(text.isEmpty())throw new IllegalArgumentException("Paste or share a message first.");
        if(text.length()>MAX_TEXT)throw new IllegalArgumentException("Keep the text under 12,000 characters.");
        Result result=new Result("message");
        for(Rule rule:rules) {
            Matcher match=rule.pattern.matcher(text);
            while(match.find()) {
                String before=text.substring(Math.max(0,match.start()-80),match.start());
                if(rule.skipNegated&&NEGATION.matcher(before).find())continue;
                add(result,rule.id,rule.severity,rule.title,rule.detail);break;
            }
        }
        if(HIDDEN.matcher(text).find())add(result,"hidden-characters","low","Hidden formatting characters","Invisible or directional characters can be legitimate, but can also disguise words or addresses.");
        List<String> links=extractLinks(text);
        for(int i=0;i<Math.min(links.size(),MAX_LINKS);i++) {
            try { Result link=checkLink(links.get(i));for(Signal signal:link.signals)add(result,"link-"+i+"-"+signal.id,signal.severity,signal.title,signal.detail); }
            catch(IllegalArgumentException exception) { add(result,"uninspected-link","medium","A link could not be inspected","Ask for a complete address or check it separately. This is not a safety verdict."); }
        }
        if(links.size()>MAX_LINKS)add(result,"link-limit","low","Some links were not inspected","Only the first 10 recognised links in a message are inspected. Check the others separately.");
        return result;
    }
    public static List<String> extractLinks(String text) {
        LinkedHashSet<String> found=new LinkedHashSet<>();Matcher matcher=LINKS.matcher(text);
        while(matcher.find()) {
            if(matcher.start()>0&&String.valueOf(text.charAt(matcher.start()-1)).matches("[@\\w]"))continue;
            String value=matcher.group().replaceAll("[,.!?;:'\\\"]+$","");
            for(int i=0;i<8&&!value.isEmpty();i++) {
                char close=value.charAt(value.length()-1);char open=close==')'?'(':close==']'?'[':close=='}'?'{':0;
                if(open==0)break;int balance=0;for(char c:value.toCharArray()) {if(c==open)balance++;if(c==close)balance--;}
                if(balance>=0)break;value=value.substring(0,value.length()-1);
            }
            if(!value.isEmpty())found.add(value);
            if(found.size()>MAX_LINKS)break;
        }
        return new ArrayList<>(found);
    }
    public Result checkLink(String input) {
        String text=input==null?"":input.trim();
        if(text.isEmpty())throw new IllegalArgumentException("Share or paste one link first.");
        if(text.length()>MAX_URL)throw new IllegalArgumentException("Use a link shorter than 4,096 characters.");
        if(Pattern.compile("\\s").matcher(text).find())throw new IllegalArgumentException("Use Message check for text containing spaces.");
        Result result=new Result("link");
        boolean scheme=Pattern.compile("^[a-z][a-z\\d+.-]*:",Pattern.CASE_INSENSITIVE).matcher(text).find()&&!text.matches("^[^/:]+:\\d+(?:[/?#].*)?$");
        result.assumedHttps=!scheme;
        URI uri;
        try { uri=new URI(scheme?text:"https://"+text); }
        catch(Exception error) { add(result,"nonstandard-url","medium","A nonstandard address","The native URL parser cannot fully interpret this address. Verify the original address independently.");return result; }
        String protocol=uri.getScheme()==null?"":uri.getScheme().toLowerCase(Locale.ROOT);
        if(!protocol.equals("http")&&!protocol.equals("https")) { add(result,"non-web-scheme","high","This is not a normal web link","This address may open another application or run an instruction. SecondLook will not open it.");return result; }
        String authority=uri.getRawAuthority();
        if(authority==null||authority.isEmpty())throw new IllegalArgumentException("The address needs a website hostname.");
        if(authority.contains("@"))add(result,"userinfo","high","Text before @ can hide the destination","Text before @ is user information, not the website. Independently verify the hostname shown below.");
        String hostPort=authority.substring(authority.lastIndexOf('@')+1);
        String host=hostPort;
        if(hostPort.startsWith("[")) { int end=hostPort.indexOf(']'); if(end<0)throw new IllegalArgumentException("Invalid IPv6 address.");host=hostPort.substring(0,end+1); }
        else if(hostPort.lastIndexOf(':')>=0)host=hostPort.substring(0,hostPort.lastIndexOf(':'));
        if(host.endsWith("."))host=host.substring(0,host.length()-1);
        boolean unicode=!StandardCharsets.US_ASCII.newEncoder().canEncode(host);
        try { if(!host.startsWith("["))host=IDN.toASCII(host,IDN.USE_STD3_ASCII_RULES); }
        catch(Exception error) { add(result,"nonstandard-host","medium","A nonstandard hostname","The hostname has unusual encoding or characters. Native and browser parsers can differ; verify it independently."); }
        host=host.toLowerCase(Locale.ROOT);result.hostname=host;
        if(host.isEmpty())throw new IllegalArgumentException("The address needs a hostname.");
        if(protocol.equals("http"))add(result,"unencrypted","medium","An unencrypted connection","HTTP does not encrypt the connection. HTTPS is preferable but is not proof that a site is trustworthy.");
        if(host.matches("(?:0x[0-9a-f]+|\\d+)(?:\\.(?:0x[0-9a-f]+|\\d+)){0,3}")||host.startsWith("["))add(result,"ip-host","medium","A numeric address instead of a domain","Numeric addresses may be legitimate, but make the organisation harder to identify. A browser may normalise them differently.");
        if(host.equals("localhost")||host.endsWith(".localhost")||host.endsWith(".local"))add(result,"local-host","low","A local-device address","Use it only if you know which local service you intend to access.");
        if(unicode||host.contains("xn--"))add(result,"international-domain","low","An internationalised domain name","These domains are often legitimate. Similar-looking letters can also be misleading. Native and browser normalisation can differ.");
        if(HIDDEN.matcher(text).find())add(result,"hidden-characters","medium","Invisible or directional characters","Hidden characters can change how an address appears. Check the original destination carefully.");
        for(String shortener:shorteners)if(belongs(host,shortener)){add(result,"shortener","low","The final destination is hidden","This offline check cannot follow a link shortener. Ask for the full original address.");break;}
        for(int i=0;i<brands.length();i++) {
            JSONObject brand=brands.optJSONObject(i);String token=brand.optString("token");
            if(!Pattern.compile("(^|[.-])"+Pattern.quote(token)+"([.-]|$)").matcher(host).find())continue;
            boolean expected=false;JSONArray domains=brand.optJSONArray("domains");for(int j=0;j<domains.length();j++)if(belongs(host,domains.optString(j)))expected=true;
            if(!expected)add(result,"brand-"+token,"medium","Brand wording outside a usual domain","The hostname mentions "+brand.optString("name")+" but is outside our small, non-exhaustive domain reference list. Open the service independently.");
        }
        String path=uri.getPath()==null?"":uri.getPath();
        if(Pattern.compile("\\.(?:exe|msi|apk|scr|bat|cmd|ps1|vbs|dmg|pkg)(?:$|/)",Pattern.CASE_INSENSITIVE).matcher(path).find())add(result,"download","high","A link that looks like a software download","No file was downloaded or inspected. Only install software from a source you independently trust.");
        String query=uri.getRawQuery();
        if(query!=null)for(String pair:query.split("&")) {
            String[] parts=pair.split("=",2);if(parts.length!=2||!Arrays.asList("url","redirect","redirect_uri","redirect_url","next","continue","destination","target").contains(parts[0]))continue;
            try { String nested=URLDecoder.decode(parts[1],"UTF-8");URI destination=new URI(nested);if(destination.getHost()!=null&&!destination.getHost().equalsIgnoreCase(host)) {add(result,"possible-redirect","low","Another website appears inside the link","A parameter names a different website. We do not know whether the server will redirect there.");break;} } catch(Exception ignored) {}
        }
        return result;
    }
    public static String nextStep(Result result) {
        for(Signal signal:result.signals)if(signal.id.contains("mistaken-deposit"))return "Check your real balance in the official banking or mobile-money app. Ask the provider to handle any reversal. Do not send money to another number.";
        return "Open the organisation’s official app or call a number you already know. Do not click, pay, share secrets, or install software until the request is independently verified. If you already acted, open the playbook.";
    }
}

File: companions/ios/Shared/SafetyEngine.swift
import Foundation

struct SafetySignal: Identifiable {
    let id: String
    let severity: String
    let title: String
    let detail: String
}
struct SafetyResult {
    let kind: String
    var hostname = ""
    var assumedHTTPS = false
    var signals: [SafetySignal] = []
    var level: String {
        if signals.contains(where: { $0.severity == "high" }) || signals.filter({ $0.severity == "medium" }).count >= 3 { return "high" }
        return signals.isEmpty ? "unknown" : "caution"
    }
    var label: String { level == "high" ? "Strong warning signs" : level == "caution" ? "Worth a closer look" : "No common warning signs found" }
    var explanation: String {
        switch level {
        case "high": return "Pause before clicking, paying, or sharing details. Independently verify the request. These patterns are not proof of a scam."
        case "caution": return "Some details deserve a closer look, but can also appear in legitimate content. Verify the context using a trusted channel."
        default: return "The known patterns were not found. This does not mean the message or link is safe. Verify unexpected requests before acting."
        }
    }
}
struct SafetyGuide: Decodable, Identifiable { let id, label, title, intro: String; let steps: [String] }
private struct MessageRule: Decodable { let id, severity, pattern, title, detail: String; let skipNegated: Bool }
private struct SafetyBrand: Decodable { let name, token: String; let domains: [String] }
private struct RuleFile: Decodable { let messageRules: [MessageRule]; let shorteners: [String]; let brands: [SafetyBrand]; let playbook: [SafetyGuide] }
enum SafetyError: LocalizedError {
    case message(String)
    var errorDescription: String? { if case .message(let value) = self { return value }; return nil }
}

/// Bounded, offline checks. This class makes no network requests and stores no input.
final class SafetyEngine {
    private let source: RuleFile
    private let compiled: [(MessageRule, NSRegularExpression)]
    var playbook: [SafetyGuide] { source.playbook }
    init(bundle: Bundle = .main) throws {
        guard let url = bundle.url(forResource: "rules", withExtension: "json") else { throw SafetyError.message("Offline rules were not found in this build.") }
        let decoded = try JSONDecoder().decode(RuleFile.self, from: Data(contentsOf: url))
        source = decoded
        compiled = try decoded.messageRules.map { ($0, try NSRegularExpression(pattern: $0.pattern, options: .caseInsensitive)) }
    }
    private func matches(_ pattern: String, _ text: String) -> Bool {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive) else { return false }
        return regex.firstMatch(in: text, range: NSRange(text.startIndex..., in: text)) != nil
    }
    private func belongs(_ host: String, _ domain: String) -> Bool { host == domain || host.hasSuffix("." + domain) }
    private func signal(_ id: String, _ severity: String, _ title: String, _ detail: String) -> SafetySignal { SafetySignal(id: id, severity: severity, title: title, detail: detail) }
    func checkMessage(_ input: String) throws -> SafetyResult {
        let text = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { throw SafetyError.message("Share or paste a message first.") }
        guard text.utf16.count <= 12000 else { throw SafetyError.message("Keep the message under 12,000 characters.") }
        var result = SafetyResult(kind: "message")
        for (rule, regex) in compiled {
            for match in regex.matches(in: text, range: NSRange(text.startIndex..., in: text)) {
                let before = (text as NSString).substring(with: NSRange(location: max(0, match.range.location - 80), length: min(80, match.range.location)))
                if rule.skipNegated && matches(#"\b(?:never|do\s+not|don['’]t|should\s+not|shouldn['’]t|will\s+not|won['’]t|avoid)\b[^,;.!?\n]{0,65}$"#, before) { continue }
                result.signals.append(signal(rule.id, rule.severity, rule.title, rule.detail)); break
            }
        }
        if matches(#"[\u200B-\u200F\u202A-\u202E\u2066-\u2069\uFEFF]"#, text) { result.signals.append(signal("hidden-characters", "low", "Hidden formatting characters", "These characters can be legitimate, but can also disguise words or addresses.")) }
        let links = extractLinks(text)
        for (index, link) in links.prefix(10).enumerated() {
            if let inspected = try? checkLink(link) {
                result.signals += inspected.signals.map { signal("link-\(index)-" + $0.id, $0.severity, $0.title, $0.detail) }
            } else { result.signals.append(signal("uninspected-link", "medium", "A link could not be inspected", "Ask for the complete address or check it separately. This is not a safety verdict.")) }
        }
        if links.count > 10 { result.signals.append(signal("link-limit", "low", "Some links were not inspected", "Only the first 10 recognised links were checked. Inspect the others separately.")) }
        return result
    }
    func extractLinks(_ text: String) -> [String] {
        let pattern = #"(?:https?://|ftp://|javascript:|data:|file:///|www\.)[^\s<>"`]+|\b(?:[a-z\d](?:[a-z\d-]{0,61}[a-z\d])?\.)+[a-z]{2,63}(?:[/:?#][^\s<>"`]*)?"#
        guard let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive) else { return [] }
        var links: [String] = []
        for match in regex.matches(in: text, range: NSRange(text.startIndex..., in: text)) {
            if match.range.location > 0 {
                let previous = (text as NSString).substring(with: NSRange(location: match.range.location - 1, length: 1))
                if matches(#"[@\w]"#, previous) { continue }
            }
            guard let range = Range(match.range, in: text) else { continue }
            var value = String(text[range])
            while let last = value.last, ",.!?;:'\"".contains(last) { value.removeLast() }
            for _ in 0..<8 {
                guard let close = value.last, let open = [Character(")"): Character("("), Character("]"): Character("["), Character("}"): Character("{")][close] else { break }
                if value.filter({ $0 == close }).count <= value.filter({ $0 == open }).count { break }
                value.removeLast()
            }
            if !value.isEmpty && !links.contains(value) { links.append(value) }
            if links.count > 10 { break }
        }
        return links
    }
    func checkLink(_ input: String) throws -> SafetyResult {
        let text = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { throw SafetyError.message("Share or paste one link first.") }
        guard text.utf16.count <= 4096 else { throw SafetyError.message("Use a link shorter than 4,096 characters.") }
        guard !matches(#"\s"#, text) else { throw SafetyError.message("Use Message check for text containing spaces.") }
        let explicitScheme = matches(#"^[a-z][a-z\d+.-]*:"#, text) && !matches(#"^[^/:]+:\d+(?:[/?#]|$)"#, text)
        var result = SafetyResult(kind: "link"); result.assumedHTTPS = !explicitScheme
        let candidate = explicitScheme ? text : "https://" + text
        guard let components = URLComponents(string: candidate) else {
            result.signals.append(signal("nonstandard-url", "medium", "A nonstandard address", "The native parser cannot fully interpret this address. Check the original independently.")); return result
        }
        let scheme = (components.scheme ?? "").lowercased()
        if !["http", "https"].contains(scheme) {
            result.signals.append(signal("non-web-scheme", "high", "This is not a normal web link", "It may open another application or run an instruction. SecondLook will not open it.")); return result
        }
        guard var host = components.host?.lowercased(), !host.isEmpty else { throw SafetyError.message("The link needs a website hostname.") }
        if host.hasSuffix(".") { host.removeLast() }; result.hostname = host
        if scheme == "http" { result.signals.append(signal("unencrypted", "medium", "An unencrypted connection", "HTTP does not encrypt the connection. HTTPS is preferable, but is not proof of trustworthiness.")) }
        if !(components.user ?? "").isEmpty || !(components.password ?? "").isEmpty {
            result.signals.append(signal("userinfo", "high", "Text before @ can hide the destination", "Text before @ is user information, not the website. Verify the parsed hostname independently."))
        }
        if matches(#"^(?:0x[0-9a-f]+|\d+)(?:\.(?:0x[0-9a-f]+|\d+)){0,3}$"#, host) || host.contains(":") {
            result.signals.append(signal("ip-host", "medium", "A numeric address instead of a domain", "It may be legitimate, but makes the organisation harder to identify. Browser and native normalisation can differ."))
        }
        if host == "localhost" || host.hasSuffix(".localhost") || host.hasSuffix(".local") { result.signals.append(signal("local-host", "low", "A local-device address", "Use it only if you know which local service you intend to access.")) }
        if host.contains("xn--") || host.unicodeScalars.contains(where: { !$0.isASCII }) { result.signals.append(signal("international-domain", "low", "An internationalised domain name", "These domains are often legitimate, but lookalike letters can be misleading. Native and browser normalisation can differ.")) }
        if matches(#"[\u200B-\u200F\u202A-\u202E\u2066-\u2069\uFEFF]"#, text) { result.signals.append(signal("hidden-characters", "medium", "Invisible or directional characters", "Hidden characters can change how the address appears. Verify the destination independently.")) }
        if source.shorteners.contains(where: { belongs(host, $0) }) { result.signals.append(signal("shortener", "low", "The final destination is hidden", "This offline check cannot follow a shortener. Ask for the complete original address.")) }
        for brand in source.brands {
            if matches("(^|[.-])" + NSRegularExpression.escapedPattern(for: brand.token) + "([.-]|$)", host) && !brand.domains.contains(where: { belongs(host, $0) }) {
                result.signals.append(signal("brand-" + brand.token, "medium", "Brand wording outside a usual domain", "The address mentions \(brand.name), but is outside our small, non-exhaustive reference list. Open the service independently."))
            }
        }
        if matches(#"\.(?:exe|msi|apk|scr|bat|cmd|ps1|vbs|dmg|pkg)(?:$|/)"#, components.path) { result.signals.append(signal("download", "high", "A link that looks like a software download", "No file was downloaded or inspected. Only install software from independently trusted sources.")) }
        for item in components.queryItems ?? [] {
            if ["url", "redirect", "redirect_uri", "redirect_url", "next", "continue", "destination", "target"].contains(item.name), let value = item.value, let embedded = URLComponents(string: value)?.host, embedded.lowercased() != host {
                result.signals.append(signal("possible-redirect", "low", "Another website appears inside the link", "A parameter names another website. We cannot determine whether the server redirects there.")); break
            }
        }
        return result
    }
}
